File: Chap#1__Introduction/Exercises/2__sum1.hs
-- Exercise#2
-- Product of a list of numbers.
-- Show that sum[x] = x for any number x.
-- Exercise on page#13: 1.7.2 -- PIH, Second Edition.
-- Prashanth Babu V V :: 01st Oct, 2016.
sum1 :: Num a => [a] -> a
sum1 []     = 0
sum1 (x:xs) = x + sum1 xs

File: Chap#1__Introduction/Exercises/2__sum1.scala
/**
 * Sum of numbers
 * Program on page#9 -- PIH, Second Edition
 * Prashanth Babu V V :: 01st Oct, 2016.
 */

object Sum1 {
  def main(args: Array[String]) {
    callMethod(List(1,2,3,4,5,6))
  }

  def callMethod(array: List[Int]): Int = {
    array match {
        case Nil => 0
        case head :: tail => head + callMethod(tail)
    }
  }
}